test/libsolidity/SolidityEndToEndTest.cpp

```diff
@@ -6853,6 +6858,64 @@ BOOST_AUTO_TEST_CASE(create_dynamic_array_with_zero_length)
 	BOOST_CHECK(callContractFunction("f()") == encodeArgs(u256(7)));
 }
 
+
+BOOST_AUTO_TEST_CASE(payable_accept_explicit_constructor)
+{
+	char const* sourceCode = R"(
+		contract C {
+			function () payable { }
+		}
+	)";
+	compileAndRun(sourceCode, 27, "C");
+}
+
+
+BOOST_AUTO_TEST_CASE(payable_accept_explicit)
+{
+	char const* sourceCode = R"(
+		contract C {
+			function f() payable returns (uint) {
+				return msg.value;
+			}
+			function() payable returns (uint) {
+				return msg.value;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "C");
+	BOOST_CHECK(callContractFunctionWithValue("f()", 27) == encodeArgs(u256(27)));
+	BOOST_CHECK(callContractFunctionWithValue("", 27) == encodeArgs(u256(27)));
+}
+
+BOOST_AUTO_TEST_CASE(non_payable_throw_constructor)
+{
+	char const* sourceCode = R"(
+		contract C {
+			function() { }
+		}
+	)";
+	compileAndRun(sourceCode, 27, "C");
```

Comment:
Ah sorry, I thought you were testing the fallback function. Why don't you use an empty contract?